core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1252,6 +1254,34 @@ private[spark] class DAGScheduler(
     execCores.map(cores => properties.setProperty(EXECUTOR_CORES_LOCAL_PROPERTY, cores))
   }
 
+  /**
+   * If push based shuffle is enabled, set the shuffle services to be used for the given
+   * shuffle map stage. The list of shuffle services is determined based on the list of
+   * active executors tracked by block manager master at the start of the stage.
+   */
+  private def prepareShuffleServicesForShuffleMapStage(stage: ShuffleMapStage) {
+    if (stage.shuffleDep.shuffleMergeEnabled) {
```

Comment:
Can you comment here in which cases this is going to be false?
From just this patch, it seems that this will always be either true/false for all the stages because it takes the value  push-based shuffle conf.

Reply:
Yes, since we don't have the finalize changes this code doesn't feel right. Technically I think we also need to have `shuffleMergeFinalized` changes which cannot be added due to missing `finalize` changes. This condition if you check in our internal branch, its basically `if (stage.shuffleDep.shuffleMergeEnabled && !stage.shuffleDep.shuffleMergeFinalized)` then check for merger locations and things like that else which means already `shuffleMergeFinalized` for this dependency so disable shuffle merge for the subsequent stages. Currently we don't want to enable shuffle merge for retry/reuse cases as it needs a complete refactor to handle all the cases.

Reply:
I would then suggest to remove this condition. I understand it is a hassle but we need this change to start the work on the shuffle write side. The finalization part in DAG scheduler will also involve a lot of discussion because there is a question of where the finalization part will go. It seems that this should be part of the change which is related to driver finalizing the shuffle. You can add a TODO.

Reply:
Like you mentioned added a TODO for now.